### ISSUE TITLE
Data copy from CPU to GPU use default stream in nightly version.

### ISSUE THREAD

**Issue by cicirori** [pytorch/pytorch#49819]
## 🐛 Bug

After https://github.com/pytorch/pytorch/pull/46304, data copy per iteration from cpu to gpu can't overlap with last iteration's computation/communication.

There is my test result which shows different behaviors by literally same code.

nightly:
![image](https://user-images.githubusercontent.com/32845984/103058754-1908de80-45de-11eb-8111-94a8d1ae81b4.png)

1.5.1 release:
![image](https://user-images.githubusercontent.com/32845984/103058787-2e7e0880-45de-11eb-9a71-48b74d2f6dbd.png)

## To Reproduce

1. Run the same code in nightly build and 1.5.1 release build. Use nsys profiling it.

My test code is listed as below, you can use it do some tests. But I think the reason that incurs different behavior is clear. Is it a new feature? If so, I might not think it's a good one.

```
import os
import time
import sys
import random
import traceback
import numpy as np
import torch

import torch.nn as nn
import torch.nn.parallel
import torch.distributed as dist
import torch.optim
import torch.utils.data
import torch.utils.data.distributed
import torchvision.transforms as transforms
import torchvision.datasets as datasets
import torchvision.models as models

from torch.multiprocessing import Pool, Process

import argparse


class AverageMeter(object):
    """Computes and stores the average and current value"""

    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count


def accuracy(output, target, topk=(1,)):
    """Computes the precision@k for the specified values of k"""
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].contiguous().view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res


def train(train_loader, model, criterion, optimizer, epoch, batch_size):
    batch_time = AverageMeter()
    data_time = AverageMeter()
    losses = AverageMeter()
    top1 = AverageMeter()
    top5 = AverageMeter()
    local_rank = os.environ['LOCAL_RANK']
    # switch to train mode
    model.train()

    end = time.time()
    for i, (input, target) in enumerate(train_loader):
        # measure data loading time
        data_time.update(time.time() - end)

        # Create non_blocking tensors for distributed training
        # input = input.cuda(non_blocking=True)
        target = target.cuda(non_blocking=True)

        # compute output
        output = model(input)
        loss = criterion(output, target)

        # measure accuracy and record loss
        prec1, prec5 = accuracy(output, target, topk=(1, 5))
        losses.update(loss.item(), input.size(0))
        top1.update(prec1[0], input.size(0))
        top5.update(prec5[0], input.size(0))

        # compute gradients in a backward pass
        optimizer.zero_grad()
        loss.backward()

        # Call step of optimizer to update model params
        optimizer.step()

        # measure elapsed time
        if i >= 10:
            batch_time.update(time.time() - end)
        end = time.time()

        if local_rank == '0' and i % 10 == 0 and i > 10:
        # if local_rank == '0' and i > 0:
            print('Epoch: [{0}][{1}/{2}]\t'
                    'Speed {speed_now:.3f} ({speed_avg:.3f})\t'
                    'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                    'Data {data_time.val:.3f} ({data_time.avg:.3f})\t'
                    'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                    'Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t'
                    'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
                        epoch, i, len(train_loader), batch_time=batch_time,
                        speed_now=batch_size/batch_time.val, speed_avg=batch_size/batch_time.avg,
                        data_time=data_time, loss=losses, top1=top1, top5=top5))


def adjust_learning_rate(initial_lr, optimizer, epoch):
    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
    lr = initial_lr * (0.1 ** (epoch // 30))
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr


def validate(val_loader, model, criterion):
    batch_time = AverageMeter()
    losses = AverageMeter()
    top1 = AverageMeter()
    top5 = AverageMeter()

    # switch to evaluate mode
    model.eval()

    with torch.no_grad():
        end = time.time()
        for i, (input, target) in enumerate(val_loader):

            input = input.cuda(non_blocking=True)
            target = target.cuda(non_blocking=True)

            # compute output
            output = model(input)
            loss = criterion(output, target)

            # measure accuracy and record loss
            prec1, prec5 = accuracy(output, target, topk=(1, 5))
            losses.update(loss.item(), input.size(0))
            top1.update(prec1[0], input.size(0))
            top5.update(prec5[0], input.size(0))

            # measure elapsed time
            torch.cuda.synchronize()
            batch_time.update(time.time() - end)
            end = time.time()

            if i % 100 == 0:
                print('Test: [{0}/{1}]\t'
                      'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                      'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                      'Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t'
                      'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
                          i, len(val_loader), batch_time=batch_time, loss=losses,
                          top1=top1, top5=top5))

        print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1, top5=top5))

    return top1.avg


def start(backend):
    print("Collect Inputs...")
    # Batch Size for training and testing
    batch_size = 64
    # Number of additional worker processes for dataloading
    workers = 8
    # Number of epochs to train for
    num_epochs = 1
    # Starting Learning Rate
    starting_lr = 0.01
    # Distributed backend type
    # dist_backend = 'nccl'
    # dist_backend = 'esl'
    dist_backend = backend
    # Url used to setup distributed training
    dist.init_process_group(dist_backend, init_method='env://')
    print("Wordsize: ", os.environ['WORLD_SIZE'])
    local_rank = int(os.environ['LOCAL_RANK'])
    print("Local_Rank: ", local_rank)

    dp_device_ids = [local_rank]
    torch.cuda.set_device(local_rank)
    device = torch.device("cuda", local_rank)

    print("Initialize Model...")
    # Construct Model
    model = models.vgg16(pretrained=False).to(device)
    # Make model DistributedDataParallel
    model = torch.nn.parallel.DistributedDataParallel(
        model, device_ids=dp_device_ids, output_device=local_rank)

    # define loss function (criterion) and optimizer
    criterion = nn.CrossEntropyLoss().cuda()
    optimizer = torch.optim.SGD(
        model.parameters(), starting_lr, momentum=0.9, weight_decay=1e-4)

    print("Initialize Dataloaders...")
    # Define the transform for the data. Notice, we must resize to 224x224 with this dataset and model.
    transform = transforms.Compose(
        [transforms.Resize(224),
         transforms.ToTensor(),
         transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

    # Initialize Datasets. STL10 will automatically download if not present
    # trainset = datasets.STL10(
    #     root='/workspace/dataset', split='train', download=True, transform=transform)
    # valset = datasets.STL10(root='/workspace/dataset',
    #                         split='test', download=True, transform=transform)
    trainset = datasets.FakeData(size=batch_size*int(os.environ['WORLD_SIZE'])*100, num_classes=1000, ,transform=transform)
    valset = datasets.FakeData(size=1000,  transform=transform)

    # Create DistributedSampler to handle distributing the dataset across nodes when training
    # This can only be called after torch.distributed.init_process_group is called
    train_sampler = torch.utils.data.distributed.DistributedSampler(trainset)

    # Create the Dataloaders to feed data to the training and validation steps
    train_loader = torch.utils.data.DataLoader(trainset, batch_size=batch_size, shuffle=(train_sampler is None),
                                               num_workers=workers, pin_memory=True, sampler=train_sampler)
    val_loader = torch.utils.data.DataLoader(valset, batch_size=batch_size, shuffle=False, num_workers=workers,
                                             pin_memory=True)

    best_prec1 = 0

    for epoch in range(num_epochs):
        # Set epoch count for DistributedSampler
        train_sampler.set_epoch(epoch)

        # Adjust learning rate according to schedule
        adjust_learning_rate(starting_lr, optimizer, epoch)

        # train for one epoch
        print("\nBegin Training Epoch {}".format(epoch + 1))
        train(train_loader, model, criterion, optimizer, epoch, batch_size*int(os.environ['WORLD_SIZE']))

        # # evaluate on validation set
        # print("Begin Validation @ Epoch {}".format(epoch + 1))
        # prec1 = validate(val_loader, model, criterion)

        # # remember best prec@1 and save checkpoint if desired
        # # is_best = prec1 > best_prec1
        # best_prec1 = max(prec1, best_prec1)

        # print("Epoch Summary: ")
        # print("\tEpoch Accuracy: {}".format(prec1))
        # print("\tBest Accuracy: {}".format(best_prec1))

def set_seed(self, seed=0):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed) #为CPU设置种子用于生成随机数，以使得结果是确定的
    torch.cuda.manual_seed(seed) #为当前GPU设置随机种子；
    torch.cuda.manual_seed_all(seed) #如果使用多个GPU，为所有的GPU设置种子。
    torch.backends.cudnn.deterministic = True # CPU和GPU结果一致
    torch.backends.cudnn.benchmark = False


if __name__ == '__main__':
    # set_seed(42)
    parser = argparse.ArgumentParser()
    parser.add_argument("--local_rank", type=int)
    args = parser.parse_args()
    os.environ['LOCAL_RANK'] = str(args.local_rank)
    backend = nccl
    start(backend)

# usage: python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_YOU_HAVE --nnodes=2 --node_rank=1 --master_addr="127.0.0.1" --master_port=1997 YOUR_TRAINING_SCRIPT.py (--arg1 --arg2 --arg3 and all other arguments of your training script)
```

## Environment

```
PyTorch version: 1.5.1                                                                                                                                │12:02:37 eth0           600B/s 0B/s
Is debug build: No                                                                                                                                     │12:02:37 eth0           600B/s 0B/s
CUDA used to build PyTorch: 10.0                                                                                                                       │12:02:37 eth0           1.17KB/s 14.02KB/s
                                                                                                                                                       │12:02:38 eth0           600B/s 0B/s
OS: Ubuntu 16.04.6 LTS                                                                                                                                 │12:02:38 eth0           600B/s 0B/s
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609                                                                                           │12:02:38 eth0           600B/s 0B/s
CMake version: version 3.5.1                                                                                                                           │12:02:38 eth0           600B/s 0B/s
                                                                                                                                                       │12:02:38 eth0           600B/s 0B/s
Python version: 3.7                                                                                                                                    │12:02:38 eth0           600B/s 0B/s
Is CUDA available: Yes                                                                                                                                 │12:02:38 eth0           600B/s 0B/s
CUDA runtime version: 10.0.130                                                                                                                         │12:02:38 eth0           600B/s 0B/s
GPU models and configuration:                                                                                                                          │12:02:39 eth0           600B/s 0B/s
GPU 0: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           600B/s 0B/s
GPU 1: Tesla V100-SXM2-16GB                                                                                                                            │time    Interface        RX        TX
GPU 2: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           600B/s 0B/s
GPU 3: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           1.17KB/s 0B/s
GPU 4: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           600B/s 0B/s
GPU 5: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           1.17KB/s 0B/s
GPU 6: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           600B/s 0B/s
GPU 7: Tesla V100-SXM2-16GB                                                                                                                            │12:02:39 eth0           600B/s 0B/s
                                                                                                                                                       │12:02:39 eth0           600B/s 0B/s
Nvidia driver version: 418.74                                                                                                                          │12:02:40 eth0           600B/s 0B/s
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5                                                                                             │12:02:40 eth0           600B/s 0B/s
                                                                                                                                                       │12:02:40 eth0           600B/s 0B/s
Versions of relevant libraries:                                                                                                                        │12:02:40 eth0           600B/s 0B/s
[pip3] numpy==1.19.2                                                                                                                                   │12:02:40 eth0           600B/s 0B/s
[pip3] torch==1.5.1                                                                                                                                    │12:02:40 eth0           600B/s 0B/s
[pip3] torchsummary==1.5.1                                                                                                                             │12:02:40 eth0           600B/s 0B/s
[pip3] torchvision==0.6.1                                                                                                                              │12:02:40 eth0           600B/s 0B/s
[conda] blas                      1.0                         mkl    defaults                                                                          │12:02:41 eth0           600B/s 0B/s
[conda] mkl                       2020.2                      256    defaults                                                                          │12:02:41 eth0           600B/s 0B/s
[conda] mkl-service               2.3.0            py37he904b0f_0    defaults                                                                          │12:02:41 eth0           600B/s 0B/s
[conda] mkl_fft                   1.2.0            py37h23d657b_0    defaults                                                                          │12:02:41 eth0           600B/s 0B/s
[conda] mkl_random                1.1.1            py37h0573a6f_0    defaults                                                                          │12:02:41 eth0           600B/s 0B/s
[conda] torch                     1.5.1                    pypi_0    pypi                                                                              │time    Interface        RX        TX
[conda] torchsummary              1.5.1                    pypi_0    pypi                                                                              │12:02:41 eth0           600B/s 0B/s
[conda] torchvision               0.6.1                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

**Comment by ngimel** [comment]
Did you verify that #46304 is causing regression? cc @rohan-varma

**Comment by rohan-varma** [comment]
Hi @cicirori, I have put up a fix at https://github.com/pytorch/pytorch/pull/50180 that hopefully resolves the regression. Feel free to patch it and give it a try and ensure communication/computation is overlapped as expected.

**Comment by cicirori** [comment]
Thanks for your quick response. I'm not familiar with how git's issue mechanism works. So should I close this issue?

**Comment by rohan-varma** [comment]
Yes, this should be fixed in master and will be part of 1.8. Will close this out
